compiler/rustc_typeck/src/check/generator_interior.rs

```diff
@@ -536,22 +536,28 @@ pub fn check_must_not_suspend_ty<'tcx>(
             }
             has_emitted
         }
-        ty::Tuple(ref tys) => {
+        ty::Tuple(_) => {
             let mut has_emitted = false;
-            let spans = if let Some(hir::ExprKind::Tup(comps)) = data.expr.map(|e| &e.kind) {
-                debug_assert_eq!(comps.len(), tys.len());
-                comps.iter().map(|e| e.span).collect()
+            let (spans, exprs) = if let Some(hir::ExprKind::Tup(comps)) = data.expr.map(|e| &e.kind)
+            {
+                debug_assert_eq!(comps.len(), ty.tuple_fields().count());
+                (comps.iter().map(|e| e.span).collect(), comps.iter().collect())
```

Comment:
I think this can avoid collecting into a vector of spans, since later you can access it by `exprs.get(i).map(|e| e.span)`, right?

Reply:
Right. Fixed.